File: Java/프로그래머스/3/388354. 홀짝트리/홀짝트리.java
import java.util.*;

public class Solution {
    public int[] solution(int[] nodes, int[][] edges) {
        // 그래프 인접 리스트
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int n : nodes) {
            graph.put(n, new ArrayList<>()); // isolated node 보장
        }
        for (int[] e : edges) {
            int a = e[0], b = e[1];
            graph.get(a).add(b);
            graph.get(b).add(a);
        }

        Set<Integer> visited = new HashSet<>();
        int ansHol = 0;
        int ansRev = 0;

        for (int start : nodes) {
            if (visited.contains(start)) continue;

            // BFS로 컴포넌트 수집
            List<Integer> comp = new ArrayList<>();
            Queue<Integer> q = new ArrayDeque<>();
            q.add(start);
            visited.add(start);

            while (!q.isEmpty()) {
                int u = q.poll();
                comp.add(u);
                for (int v : graph.get(u)) {
                    if (!visited.contains(v)) {
                        visited.add(v);
                        q.add(v);
                    }
                }
            }

            int holNonrootCnt = 0; // 비-루트 가정에서 "홀짝" 노드 개수
            int revNonrootCnt = 0; // 비-루트 가정에서 "역홀짝" 노드 개수

            for (int u : comp) {
                int deg = graph.get(u).size();
                int nonrootChildren = deg - 1; // 비-루트일 때 자식 수
                if ((u % 2) == (nonrootChildren % 2)) {
                    holNonrootCnt++;
                } else {
                    revNonrootCnt++;
                }
            }

            if (revNonrootCnt == 1) ansHol++;
            if (holNonrootCnt == 1) ansRev++;
        }

        return new int[]{ansHol, ansRev};
    }
}

File: Python3/프로그래머스/3/388354. 홀짝트리/홀짝트리.py
from collections import defaultdict, deque

def solution(nodes, edges):
    graph = defaultdict(list)
    # 그래프 생성 (무향)
    for a, b in edges:
        graph[a].append(b)
        graph[b].append(a)
    # isolated node 도 키로 보장
    for n in nodes:
        if n not in graph:
            graph[n] = []

    visited = set()
    ans_hol = 0   # 홀짝 트리 개수
    ans_rev = 0   # 역홀짝 트리 개수

    for start in nodes:
        if start in visited:
            continue
        # 컴포넌트 수집 (BFS)
        q = deque([start])
        visited.add(start)
        comp = []
        while q:
            u = q.popleft()
            comp.append(u)
            for v in graph[u]:
                if v not in visited:
                    visited.add(v)
                    q.append(v)

        hol_nonroot_cnt = 0   # 비-루트 가정에서 '홀짝 노드'인 개수
        rev_nonroot_cnt = 0   # 비-루트 가정에서 '역홀짝 노드'인 개수

        for u in comp:
            deg = len(graph[u])
            nonroot_children = deg - 1
            # 비-루트일 때의 자식 개수의 홀짝과 노드 번호 홀짝 비교
            if (u % 2) == (nonroot_children % 2):
                hol_nonroot_cnt += 1
            else:
                rev_nonroot_cnt += 1

        # rev_nonroot_cnt == 1 이면 그 1개를 루트로 택하면 전체가 '홀짝 트리'가 됨
        if rev_nonroot_cnt == 1:
            ans_hol += 1
        # hol_nonroot_cnt == 1 이면 그 1개를 루트로 택하면 전체가 '역홀짝 트리'가 됨
        if hol_nonroot_cnt == 1:
            ans_rev += 1

    return [ans_hol, ans_rev]